guides/06_client-libraries/fastapi-app-with-the-gradio-client.md

```diff
@@ -0,0 +1,261 @@
+
+```python
+import subprocess
+
+def process_video(video_path):
+    old_audio = os.path.basename(video_path).split(".")[0] + ".m4a"
+    subprocess.run(['ffmpeg', '-y', '-i', video_path, '-vn', '-acodec', 'copy', old_audio])
+    
+    new_audio = acapellify(old_audio)
+    
+    new_video = f"acap_{video_path}"
+    subprocess.call(['ffmpeg', '-y', '-i', video_path, '-i', new_audio, '-map', '0:v', '-map', '1:a', '-c:v', 'copy', '-c:a', 'aac', '-strict', 'experimental', f"static/{new_video}"])
+    return new_video
+```
+
+You can read up on [ffmpeg documentation](https://ffmpeg.org/ffmpeg.html) if you'd like to understand all of the command line parameters, as they are beyond the scope of this tutorial.
+
+## Step 2: Create a FastAPI app (Backend Routes)
+
+Next up, we'll create a simple FastAPI app. If you haven't used FastAPI before, check out [the great FastAPI docs](https://fastapi.tiangolo.com/). Otherwise, this basic template, which we add to `main.py`, will look pretty familiar:
+
+```python
+import os
+from fastapi import FastAPI, File, UploadFile, Request
+from fastapi.responses import HTMLResponse, RedirectResponse
+from fastapi.staticfiles import StaticFiles
+from fastapi.templating import Jinja2Templates
+
+app = FastAPI()
+os.makedirs("static", exist_ok=True)
+app.mount("/static", StaticFiles(directory="static"), name="static")
+templates = Jinja2Templates(directory="templates")
+
+videos = []
+
+@app.get("/", response_class=HTMLResponse)
+async def home(request: Request):
+    return templates.TemplateResponse(
+        "home.html", {"request": request, "videos": videos})
+
+@app.post("/uploadvideo/")
+async def upload_video(video: UploadFile = File(...)):
+    new_video = process_video(video.filename)
+    videos.append(new_video)
+    return RedirectResponse(url='/', status_code=303)
+```
+
+In this example, the FastAPI app has two routes: `/` and `/uploadvideo/`.
+
+The `/` route returns an HTML template that displays a gallery of all uploaded videos. 
+
+The `/uploadvideo/` route accepts a `POST` request with an `UploadFile` object, which represents the uploaded video file. The video file is "acapellified" via the `process_video()` method, and the output video is stored in a list which stores all of the uploaded videos in memory.
+
+Note that this is a very basic example and if this were a production app, you will need to add more logic to handle file storage, user authentication, and security considerations. 
+
+## Step 3: Create a FastAPI app (Frontend Template)
+
+Finally, we create the frontend of our web application. First, we create a folder called `templates` in the same directory as `main.py`. We then create a template, `home.html` inside the `templates` folder. Here is the resulting file structure:
+
+```
+├── main.py
+├── templates
+│   └── home.html
+```
+
+Write the following as the contents of `home.html`:
+
+```html
+<!DOCTYPE html>
+<html>
```

Comment:
it's kind of annoying, but right now html written in a codeblock actually gets rendered on the page... because these guides go through a markdown to html generator. resulting in this codeblock that doesn't make sense

<img width="926" alt="Screen Shot 2023-05-02 at 7 33 38 PM" src="https://user-images.githubusercontent.com/9021060/235714105-8e6fa461-8586-4f3b-832e-4cc88687a274.png">



to get around this you shouldn't have both the open and close tags (both < and >), instead replace one with 

I did it for this codeblock: 

```
&lt;!DOCTYPE html>
&lt;html>
  &lt;head>
    &lt;title>Video Gallery&lt;/title>
    &lt;style>
      body {
        font-family: sans-serif;
        margin: 0;
        padding: 0;
        background-color: #f5f5f5;
      }
      h1 {
        text-align: center;
        margin-top: 30px;
        margin-bottom: 20px;
      }
      .gallery {
        display: flex;
        flex-wrap: wrap;
        justify-content: center;
        gap: 20px;
        padding: 20px;
      }
      .video {
        border: 2px solid #ccc;
        box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
        border-radius: 5px;
        overflow: hidden;
        width: 300px;
        margin-bottom: 20px;
      }
      .video video {
        width: 100%;
        height: 200px;
      }
      .video p {
        text-align: center;
        margin: 10px 0;
      }
      form {
        margin-top: 20px;
        text-align: center;
      }
      input[type="file"] {
        display: none;
      }
      .upload-btn {
        display: inline-block;
        background-color: #3498db;
        color: #fff;
        padding: 10px 20px;
        font-size: 16px;
        border: none;
        border-radius: 5px;
        cursor: pointer;
      }
      .upload-btn:hover {
        background-color: #2980b9;
      }
      .file-name {
        margin-left: 10px;
      }
    &lt;/style>
  &lt;/head>
  &lt;body>
    &lt;h1>Video Gallery&lt;/h1>
    {% if videos %}
      &lt;div class="gallery">
        {% for video in videos %}
          &lt;div class="video">
            &lt;video controls>
              &lt;source src="{{ url_for('static', path=video) }}" type="video/mp4">
              Your browser does not support the video tag.
            &lt;/video>
            &lt;p>{{ video }}&lt;/p>
          &lt;/div>
        {% endfor %}
      &lt;/div>
    {% else %}
      &lt;p>No videos uploaded yet.&lt;/p>
    {% endif %}
    &lt;form action="/uploadvideo/" method="post" enctype="multipart/form-data">
      &lt;label for="video-upload" class="upload-btn">Choose video file&lt;/label>
      &lt;input type="file" name="video" id="video-upload">
      &lt;span class="file-name">&lt;/span>
      &lt;button type="submit" class="upload-btn">Upload&lt;/button>
    &lt;/form>
    &lt;script>
      // Display selected file name in the form
      const fileUpload = document.getElementById("video-upload");
      const fileName = document.querySelector(".file-name");

      fileUpload.addEventListener("change", (e) => {
        fileName.textContent = e.target.files[0].name;
      });
    &lt;/script>
  &lt;/body>
&lt;/html>

```

Reply:
Thank you!